### PR TITLE
Коннов - Лабораторная работа  #2

### PR DESCRIPTION
**Pull request by SergeyKonnov**

### CODE REVIEW

**Comment by SergeyKonnov** [comment]
@gooddoog @alex_di code review pls

File: modules/tree/src/tree.cpp [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    DelAll(root);
+}
+
+void Tree::Add(int data) {
+    root = AddNode(root, data);
+}
+
+void Tree::Del(int data) {
+    if (DelNode(root, data))
+        root = nullptr;
```
Comment (gooddoog) [inline modules/tree/src/tree.cpp]: This condition should be reversed.
This test fails:
```cpp
TEST(TreeTest, search_after_del) {
    // Arrange
    int data = 1;
    Tree t;

    // Act
    t.Add(data);
    t.Add(data + 1);
    t.Del(data);
    t.Search(data + 1);

    // Assert
    EXPECT_EQ(0, t.Search(data));
    EXPECT_EQ(1, t.Search(data + 1));
}
```

File: modules/tree/test/test_main.cpp [hunk 2]
```diff
@@ -0,0 +1,8 @@
+// Copyright 2017 Korniakov Kirill
```
Comment (gooddoog) [inline modules/tree/test/test_main.cpp]: Change first name, last name and year

File: modules/tree/src/tree.cpp [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        node->left = AddNode(node->left, data);
+    } else if (data > node->data) {
+        node->right = AddNode(node->right, data);
+    } else {
+        node->count+=cnt;
+    }
+    return node;
+}
+
+Tree::Node* Tree::CopyNode(Node* node, Node* copyroot) {
```
Comment (gooddoog) [inline modules/tree/src/tree.cpp]: It's relevant to call this function "CopyTree" or "CopySubtree", because you copy all subtree, starting with some root.

Comment (SergeyKonnov) [reply modules/tree/src/tree.cpp]: fixed.

File: modules/tree/test/test_main.cpp (hunk 2 above)
Comment (SergeyKonnov) [reply modules/tree/test/test_main.cpp]: fixed

File: modules/tree/src/tree.cpp (hunk 1 above)
Comment (SergeyKonnov) [reply modules/tree/src/tree.cpp]: fixed.

**Comment by gooddoog** [comment]
:+1:

**Comment by SergeyKonnov** [comment]
@left841

**Comment by left841** [comment]
:+1:

**Comment by SergeyKonnov** [comment]
@TheG1uy @alexbatashev @Agricola87

File: modules/tree/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# Declare variables for binaries' names
+get_filename_component(DIR_NAME ${CMAKE_CURRENT_LIST_DIR} NAME)
+set(MODULE      "${DIR_NAME}")
+set(LIBRARY     "lib_${MODULE}")
+set(TESTS       "test_${MODULE}")
+set(APPLICATION "app_${MODULE}")
+
+# Include directory with public headers
+include_directories("./")
```
Comment (alexbatashev) [inline modules/tree/CMakeLists.txt]: CMAKE_CURRENT_SOURCE_DIR

File: modules/tree/test/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,27 @@
+set(target ${TESTS})
+
+file(GLOB srcs "*.cpp")
+set_source_files_properties(${srcs} PROPERTIES
+    LABELS "${MODULE};Test")
+
+# VS2012 doesn't support correctly the tuples yet,
```
Comment (alexbatashev) [inline modules/tree/test/CMakeLists.txt]: Это точно вам нужно?

File: modules/tree/CMakeLists.txt (hunk 4 above)
Comment (SergeyKonnov) [reply modules/tree/CMakeLists.txt]: Исправил.

File: modules/tree/test/CMakeLists.txt (hunk 5 above)
Comment (SergeyKonnov) [reply modules/tree/test/CMakeLists.txt]: Убрал.

**Pull request merged**